### ISSUE TITLE
Artefacts after tab-completion (OSX 10.15.3 + iTerm2 3.3.9)

### ISSUE THREAD

**Issue by killermoehre** [akinomyoga/ble.sh#42]
**ble version**: 0.4.0-devel2+2184739
**Bash version**: 5.0.11(1)-release

After tab-completion and choosing of a suggestion, the other suggestions are not cleared, but stay in the terminal. It seems like there is a line break to much or to less.
How can I debug this further?

**Comment by akinomyoga** [comment]
Thank you for the report! Currently I couldn't have reproduced the problem. 

- Does it reproduce with other terminals in OSX 10.15.3?
- Does it reproduce with an older version ble-0.3 (which is available in the git branch `v0.3-master`)?
- What is your setting of `$PS1`? Does it reproduce with a simple `PS1` such as `PS1='\$ '`?

> It seems like there is a line break to much or to less.

- After the uncleared suggestion, is the cursor position on the correct line?
  - In the case that the cursor position is wrong, is it above the expected line or below?
  - Does it also reproduce when the prompt and command line is at the top line of the terminal? (You can clear the screen and go to the top line with <kbd>C-l</kbd>. If you have line breaks in `PS1`, could you also try with a single line `PS1`?).
- Does it always reproduce with different set of suggestions? For example, does the same problem occur in other directories with many suggestions or small number of suggestions.
- What is the value of the environment variable `$TERM`?
- What is the result of the following commands?

```bash
$ cat -v <<< $_ble_term_ind
$ tput ind | cat -v
$ tput sf | cat -v
$ cat -v <<< $_ble_term_ri
$ tput ri | cat -v
$ tput sr | cat -v
$ cat -v <<< $_ble_term_dl
$ tput dl | cat -v
$ tput DL | cat -v
```

> How can I debug this further?

If you would like to dig into the codes, the clear of suggestions is made by the following function calls.

```
ble/widget/menu/accept (lib/core-complete.sh)
  ble/complete/menu#clear (lib/core-complete.sh)
    ble-edit/info/immediate-clear (src/edit.sh)
      ble-edit/info/.render-content (src/edit.sh)
        ble-edit/info/reveal (src/edit.sh)
          ble-edit/info/.clear-content (src/edit.sh)
```

I would appreciate it if you could check that the innermost function call `ble-edit/info/.clear-content` is actually called when you choose a suggestion [ Note: After you edit the files, you need to run `make` (and also `make install` if necessary) to update `ble.sh` and related files ].

**Comment by killermoehre** [comment]
> * Does it reproduce with other terminals in OSX 10.15.3?
yes
> * Does it reproduce with an older version ble-0.3 (which is available in the git branch `v0.3-master`)?
yes
> * What is your setting of `$PS1`? 
```Shell
$ echo $PS1
\[\]\[\]\[\033]0;\u@\H:\W\007\]\[$(iterm2_prompt_mark)\]\$ \[\]\[\]
```
> Does it reproduce with a simple `PS1` such as `PS1='\$ '`?
yes
> > It seems like there is a line break to much or to less.
> 
> * After the uncleared suggestions, is the cursor position on the correct line?
yes
> * Does it also reproduce when the prompt and command line is at the top line of the terminal? (You can clear the screen and go to the top line with C-l. If you have line breaks in `PS1`, could you also try with a single line `PS1`?).
yes
> * Does it always reproduce with different set of suggestions? For example, does the same problem occur in other directories with many suggestions or small number of suggestions.
no, only for multiline suggestions
> * What is the value of the environment variable `$TERM`?
```Shell
$ echo $TERM
xterm-256color
```
> * What is the result of the following commands?
> ```Shell
> $ cat -v <<< $_ble_term_ind
> ```
```Shell
$ cat -v <<< $_ble_term_ind


```
> ```Shell
> $ tput ind | cat -v
> ```
```Shell
$ tput ind | cat -v

```
> ```Shell
> $ tput sf | cat -v
> ```
```Shell
$ tput sf | cat -v

```
> ```Shell
> $ cat -v <<< $_ble_term_ri
> ```
```Shell
$ cat -v <<< $_ble_term_ri
^[M
```
> ```Shell
> $ tput ri | cat -v
> ```
```Shell
$ tput ri | cat -v
^[M[ble: EOF]                                                                                                                                                                                  
```
> ```Shell
> $ tput sr | cat -v
> ```
```Shell
$ tput sr | cat -v
^[M[ble: EOF]                                                                                                                                                                                  
```
> ```Shell
> $ cat -v <<< $_ble_term_dl
> ```
```Shell
$ cat -v <<< $_ble_term_dl
^[[M
```
> ```Shell
> $ tput dl | cat -v
> ```
```Shell
$ tput dl | cat -v
^[[M[ble: EOF]                                                                                                                                                                                 
```
> ```Shell
> $ tput DL | cat -v
> ```
```Shell
$ tput DL | cat -v
^[[%p1%dM[ble: EOF]                                                                                                                                                                            
```
> 
> > How can I debug this further?
> 
> If you would like to dig into the codes, the clear of suggestions is made by the following function calls.
> 
> ```
> ble/widget/menu/accept (lib/core-complete.sh)
>   ble/complete/menu#clear (lib/core-complete.sh)
>     ble-edit/info/immediate-clear (src/edit.sh)
>       ble-edit/info/.render-content (src/edit.sh)
>         ble-edit/info/reveal (src/edit.sh)
>           ble-edit/info/.clear-content (src/edit.sh)
> ```
> 
> I would appreciate it if you could check that the innermost function call `ble-edit/info/.clear-content` is actually called when you choose a suggestion [ Note: After you edit the files, you need to run `make` (and also `make install` if necessary) to update `ble.sh` and related files ].
> 
> **Edit** If you have any other things that you noticed, please feel free to write it here.
Maybe I find some time to dig into this. Thanks for your help

**Comment by akinomyoga** [comment]
Thank you very much for your detailed results! I think I found the reason. I have pushed a fix f8c54ef to master branch. Could you try it with `ble-update` to see if the problem is fixed?

----

**Summary**: ble.sh uses `tput` to cache terminfo/termcap entries which will be used to render command line contents. Depending on the system, `tput` supports either or both of terminfo entry name and termcap entry name. When both terminfo/termcap is supported, ble.sh uses terminfo names to access entries. But there are collisions of entry names between terminfo and termcap, i.e., some different features share the same name. One example is `dl` which means "Delete *N* Lines" in terminfo and "Delete Single Line" in termcap. In fact `tput` seems to output termcap results for `dl` in your system, which contradicted with ble.sh's assumption.

>> - After the uncleared suggestions, is the cursor position on the correct line?
>
>yes

> no, only for multiline suggestions

> ```console
> $ cat -v <<< $_ble_term_dl
> ^[[M
> $ tput dl | cat -v
> ^[[M[ble: EOF]
> $ tput DL | cat -v
> ^[[%p1%dM[ble: EOF]
> ```

Thank you very much! The above results were particularly important hints to pin down the issue.

**Comment by killermoehre** [comment]
Works for me. Thank you very much.
